### PR TITLE
win: fix watching root files

### PR DESCRIPTION
**Pull request by seishun**
When passing `\\?\C:` to CreateFile, it opens the drive rather than the root directory. So include the trailing backslash in the directory name.

Fixes nodejs/node#4643

### CODE REVIEW

**Comment by saghul** [comment]
Thanks! Could you please write a test case for this?

**Comment by seishun** [comment]
I'm not sure how I would test something like that. It requires a certain file to be in the root of a certain drive. Creating a file in the root directory usually requires administrator privileges.

**Comment by saghul** [comment]
Maybe you can try to read a file known to exist on a default windows installation? And skip the test if it cannot be found.

**Comment by seishun** [comment]
Are there any tests which are specifically for Windows?

**Comment by saghul** [comment]
Many, check the idefs in test-list.h.

**Comment by seishun** [comment]
Added a test case.

File: test/test-fs-event.c [hunk 1]
```diff
@@ -109,6 +109,12 @@ static void fail_cb(uv_fs_event_t* handle,
   ASSERT(0 && "fail_cb called");
 }
 
+static void noop_cb(uv_fs_event_t* handle,
+                    const char* path,
+                    int events,
+                    int status) {
+}
```
Comment (saghul) [inline test/test-fs-event.c]: abort here, since the callback should never be called.

File: test/test-fs-event.c [hunk 2]
```diff
@@ -109,6 +109,12 @@ static void fail_cb(uv_fs_event_t* handle,
   ASSERT(0 && "fail_cb called");
 }
 
+static void noop_cb(uv_fs_event_t* handle,
```
Comment (saghul) [inline test/test-fs-event.c]: ifdef this

File: test/test-fs-event.c [hunk 3]
```diff
@@ -531,6 +537,32 @@ TEST_IMPL(fs_event_watch_file_current_dir) {
   return 0;
 }
 
+TEST_IMPL(fs_event_watch_file_root_dir) {
```
Comment (saghul) [inline test/test-fs-event.c]: ifdef the test

File: test/test-fs-event.c [hunk 4]
```diff
@@ -531,6 +537,32 @@ TEST_IMPL(fs_event_watch_file_current_dir) {
...
+  uv_loop_t* loop;
+  int r;
+
+  const char* sys_drive = getenv("SystemDrive");
```
Comment (saghul) [inline test/test-fs-event.c]: assert that it's not NULL

File: test/test-fs-event.c [hunk 5]
```diff
@@ -531,6 +537,32 @@ TEST_IMPL(fs_event_watch_file_current_dir) {
...
+  char path[] = "\\\\?\\X:\\autoexec.bat";
+  strncpy(path + sizeof("\\\\?\\") - 1, sys_drive, 1);
+
+  loop = uv_default_loop();
+
+  r = uv_fs_event_init(loop, &fs_event);
+  ASSERT(r == 0);
+  r = uv_fs_event_start(&fs_event, noop_cb, path, 0);
+  if (r == UV__ENOENT) {
```
Comment (saghul) [inline test/test-fs-event.c]: UV_ENOENT

File: test/test-fs-event.c [hunk 6]
```diff
@@ -531,6 +537,32 @@ TEST_IMPL(fs_event_watch_file_current_dir) {
...
+    fprintf(stderr,
```
Comment (saghul) [inline test/test-fs-event.c]: Use RETURN_SKIP

**Comment by saghul** [comment]
Made a quick-pass review.

File: test/test-fs-event.c (hunk 1 above)
Comment (seishun) [inline test/test-fs-event.c]: You're right, we aren't even starting the loop. I guess it's better to just use fail_cb then.

Comment (saghul) [inline test/test-fs-event.c]: MAKE_VALGRIND_HAPPY will start the loop after all handles are closed in order for the close callbacks to run. This callback should never be called anyway.

File: test/test-fs-event.c (hunk 4 above)
Comment (seishun) [inline test/test-fs-event.c]: Maybe skip the test if it's NULL?

Comment (saghul) [inline test/test-fs-event.c]: since the test is running on Windows I'd rather assert.

**Comment by seishun** [comment]
Made the requested changes.

**Comment by saghul** [comment]
CI: https://ci.nodejs.org/job/libuv+any-pr+multi/227/

**Comment by seishun** [comment]
Looks like it succeeded on Windows.

**Comment by saghul** [comment]
But they were skipped :-( Maybe there is a better file which could be
watched?

**Comment by seishun** [comment]
Does anyone know what files the CI server has in the root directory?

**Comment by amurzeau** [comment]
what about pagefile.sys ? (and possibly a read of "HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Session Manager\Memory Management\ExistingPageFiles" to support system where the pagefile is not on C:\ drive)

**Comment by saghul** [comment]
@seishun IIRC, NTFS is case sensitive. Maybe try AUTOEXEC.BAT ?

**Comment by seishun** [comment]
@saghul It's case insensitive, and it's called `autoexec.bat` for me.

I'll try using the registry to get the pagefile drive, but at this point it seems to be getting too complicated for a test...

**Comment by saghul** [comment]
@orangemocha perhaps you can help :-) any suggestions about the file we
should be looking for?

**Comment by orangemocha** [comment]
I think c:\pagefile.sys is more likely to be there than autoexec.bat. You could also try a directory listing and use the first file you find. Though I am not sure if either approach is 100% guaranteed to work (eg system files, exclusive access).

**Comment by orangemocha** [comment]
or c:\bootsect.bak (verified this is present on the CI machines)

**Comment by seishun** [comment]
Replaced autoexec.bat with bootsect.bat.

**Comment by saghul** [comment]
CI is timing out on me :-S

**Comment by saghul** [comment]
I think I managed to trigger a new CI: https://ci.nodejs.org/job/libuv+any-pr+multi/235/

**Comment by saghul** [comment]
Tests passed! I'll have another look later today before giving my LGTM.

**Comment by saghul** [comment]
LGTM. I started landing it and noticed a couple of nits so I amended your commit. Landed in 3c59ad6. Thanks!

**Comment by orangemocha** [comment]
Nice! Great work @seishun and @saghul :smile:

**Pull request closed without merging**